File: src/game/data/loadDatabase.ts
import {
  assertGameDatabaseShape,
  buildIndexes,
  countNeedsData,
  filterLaunchContent,
  validateDatabase,
} from './validate'
import type { DatabaseIndexes, GameDatabase, ValidationIssue } from './types'

/** Bump when Launch content rows change so browsers skip stale JSON. */
export const DATABASE_CONTENT_VERSION = '2026-08-11-lucky-necklace-drop-chance'
export const DATABASE_URL = `/data/game-database.json?v=${DATABASE_CONTENT_VERSION}`

export interface LoadedDatabase {
  /** Full source database (never mutated / filtered away). */
  source: GameDatabase
  /** Launch-phase view for runtime content selection. */
  launch: GameDatabase
  sourceIndexes: DatabaseIndexes
  launchIndexes: DatabaseIndexes
  issues: ValidationIssue[]
  needsDataCount: number
}

export async function fetchDatabase(url: string = DATABASE_URL): Promise<unknown> {
  const response = await fetch(url, { cache: 'no-cache' })
  if (!response.ok) {
    throw new Error(`Failed to load database (${response.status}) from ${url}`)
  }
  return response.json()
}

export function prepareDatabase(raw: unknown): LoadedDatabase {
  assertGameDatabaseShape(raw)
  const source = raw
  const issues = validateDatabase(source)
  const errors = issues.filter((issue) => issue.severity === 'error')
  if (errors.length > 0) {
    const summary = errors
      .slice(0, 5)
      .map((issue) => `${issue.table ?? 'root'}: ${issue.message}`)
      .join('; ')
    throw new Error(`Database validation failed (${errors.length} error(s)): ${summary}`)
  }

  const launch = filterLaunchContent(source)
  return {
    source,
    launch,
    sourceIndexes: buildIndexes(source),
    launchIndexes: buildIndexes(launch),
    issues,
    needsDataCount: countNeedsData(source),
  }
}

export async function loadDatabase(url: string = DATABASE_URL): Promise<LoadedDatabase> {
  const raw = await fetchDatabase(url)
  return prepareDatabase(raw)
}

File: packages/ik_content/lib/src/load_database.dart
import 'generated/rows.dart';
import 'indexes.dart';
import 'validate.dart';

/// Bump when Launch content rows change, matching `DATABASE_CONTENT_VERSION` in
/// [src/game/data/loadDatabase.ts](../../../../src/game/data/loadDatabase.ts).
const String databaseContentVersion = '2026-09-03-golden-spud';

/// Path of the shared database inside `content/`. How those bytes are read is
/// the host's problem: this package stays free of IO so it can be tested
/// headlessly and reused on every platform.
const String databaseAssetPath = 'data/game-database.json';

class LoadedDatabase {
  const LoadedDatabase({
    required this.source,
    required this.launch,
    required this.sourceIndexes,
    required this.launchIndexes,
    required this.issues,
    required this.needsDataCount,
  });

  /// Full source database, never filtered.
  final GameDatabase source;

  /// Launch-phase view used for runtime content selection.
  final GameDatabase launch;

  final DatabaseIndexes sourceIndexes;
  final DatabaseIndexes launchIndexes;
  final List<ValidationIssue> issues;
  final int needsDataCount;
}

/// Validates [raw] and builds the runtime views.
///
/// Throws [DatabaseShapeException] when a table is missing and
/// [DatabaseValidationException] when validation finds errors, with the same
/// message the TypeScript loader produces.
LoadedDatabase prepareDatabase(Object? raw) {
  final source = assertGameDatabaseShape(raw);
  final issues = validateDatabase(source);
  final errors = issues.where((issue) => issue.isError).toList(growable: false);
  if (errors.isNotEmpty) {
    final summary = errors
        .take(5)
        .map((issue) => '${issue.table ?? 'root'}: ${issue.message}')
        .join('; ');
    throw DatabaseValidationException(
      'Database validation failed (${errors.length} error(s)): $summary',
      errors,
    );
  }

  final launch = filterLaunchContent(source);
  return LoadedDatabase(
    source: source,
    launch: launch,
    sourceIndexes: buildIndexes(source),
    launchIndexes: buildIndexes(launch),
    issues: issues,
    needsDataCount: countNeedsData(source),
  );
}
